### ISSUE TITLE
Sourceror.get_range/1 raises for keyword list without brackets

### ISSUE THREAD

**Issue by msaraiva** [doorgan/sourceror#26]
Hi @doorgan!

I recently run into this problem when using `get_range/1`. 

Here's a minimal example:

```elixir
alias Sourceror.Zipper, as: Z

code = 
  """
  config :my_app, :some_key,
    value: [
      a: 1
    ]
  """

last_arg_node = code |> Sourceror.parse_string!() |> Z.zip() |> Z.down |> Z.rightmost() |> Z.node()
```

Printing the node works fine:

```elixir
iex> last_arg_node |> Sourceror.to_string(format: :splicing) |> IO.puts

  value: [
    a: 1
  ]

:ok
```

So I guess I have the correct node. However, if I tried to get the range with:

```elixir
Sourceror.get_range(last_arg_node)
```

This raises:

```
** (FunctionClauseError) no function clause matching in Sourceror.Range.do_get_range/1    
    
    The following arguments were given to Sourceror.Range.do_get_range/1:
    
        # 1
        [
          {{:__block__,
            [
              trailing_comments: [],
              leading_comments: [],
              format: :keyword,
              line: 2,
              column: 3
            ], [:value]},
           {:__block__,
            [ 
              trailing_comments: [],
              leading_comments: [],
              newlines: 1,
              closing: [line: 4, column: 3],
              line: 2,
              column: 10
            ],
            [
              [
                {{:__block__,
                  [
                    trailing_comments: [],
                    leading_comments: [],
                    format: :keyword,
                    line: 3,
                    column: 5
                  ], [:a]},
                 {:__block__,
                  [
                    trailing_comments: [],
                    leading_comments: [],
                    token: "1",
                    line: 3,
                    column: 8
                  ], [1]}}
              ]
            ]}}
        ]
    
    Attempted function clauses (showing 10 out of 18):
    
        defp do_get_range({:__aliases__, meta, segments})
        defp do_get_range({:__block__, meta, [string]}) when is_binary(string)
        defp do_get_range({:__block__, meta, [number]}) when is_integer(number) or is_float(number)
        defp do_get_range({:__block__, meta, [atom]}) when is_atom(atom)
        defp do_get_range({:__block__, _, args} = quoted)
        defp do_get_range({form, meta, context}) when is_atom(form) and is_atom(context)
        defp do_get_range({left, right})
        defp do_get_range({{:., _, [Access, :get]}, _, _} = quoted)
        defp do_get_range({{:., _, [_, :{}]}, _, _} = quoted)
        defp do_get_range({{:., _, [:erlang, :binary_to_atom]}, meta, [interpolation, :utf8]})
        ...
        (8 clauses not shown)
    
    (sourceror 0.8.2) lib/sourceror/range.ex:50: Sourceror.Range.do_get_range/1
    (sourceror 0.8.2) lib/sourceror/range.ex:11: Sourceror.Range.get_range/2
```

If the keyword list is wrapped in brackets, like:

```elixir
  config :my_app, :some_key, [
    value: [
      a: 1
    ]
  ]
```

It works just fine. I'm I doing something wrong?

**Comment by doorgan** [comment]
This definitely looks like a bug, I'll take a look later today

**Comment by doorgan** [comment]
Yup, was a bug. For a syntax like `config :foo, :bar, baz: :qux`, the last value is actually `{:baz, :qux}` and not `[{:baz, :qux}]`. I wasn't accounting for that, so the function crashed because of a missing clase. I released v0.8.3 with a fix, let me know how it works c:

**Comment by doorgan** [comment]
Oh I see the issue here. Yes you're correct, there's still a bug in the implementation.

On the other hand, there's a limitation in the ranges functionality when it comes to modified trees: new nodes or modified nodes are not guaranteed to have the correct line and column information, or that it is present at all. This means that if you alter the tree, you can:
a) get incorrect numbers
b) not be able to get them at all(because they don't have such metadata)

The documentation for that function can be clearer on that. Sadly I wan't able to find a way around this limitation yet.

I have investigated the ruby [parser](https://github.com/whitequark/parser) gem after Marc-André Lafortune pointed it out to me. It has a `Rewriter` class that does something similar to what Sourceror does(it can generate patches for text ranges), but has some scaffolding to correct numbers when a patch is applied. For example, say you start with a source code and its AST, a `Rewriter` for the tree would be like an Ecto changeset: it holds both the original text and its ast. You apply functions to manipulate it and each change is stored(or "scheduled") to be run with a `Rewriter.apply`. When that happens, the patches are applied to the text and a new AST is generated from that new text, and so you're guaranteed to always have the correct position numbers. Some details are stil a bit fuzzy for me as I'm not fluent in Ruby, but at a high level that's how it works.

Something like that could be implemented in Sourceror, but in summary you need to apply the patches and reparse in order to get actual, correct line numbers.

Calculating the line numbers after a patch would be applied may also be viable: I haven't explored this yet.

**Comment by msaraiva** [comment]
> This means that if you alter the tree, you can:
> a) get incorrect numbers
> b) not be able to get them at all(because they don't have such metadata)

Yeah, that makes sense.

> Also at this stage this is more about making it possible than making it as efficient as possible :p Parsing, reparsing, traversing and retraversing is definitely not cheap in big projects

For my use cases that wouldn't be an issue as I just need to update a few files when bootstrapping a fresh project. I'll try this approach later. Maybe creating a higher lever abstraction that can do the following in a nice way:

1. Parse the code (and keep it for later use - patching!)
1. Use `zippers` to make it easier to find nodes and register patches for them
1. Apply all registered patches one by one, always re-parsing the previously patched code before applying the next patch

I'll let you know, if I can get anything done :)

Thanks!

**Comment by doorgan** [comment]
Yup, that would work :)

And no worries! Thank's for using Sourceror and for the feedback :)

PS: I'll keep this one open for a bit, it currently crashes with something like `config :foo, :bar, a: 1, b: 2`. I'll try to fix that once I get back home

**Comment by doorgan** [comment]
There was an issue with stabs like `a -> b` too because of the same bug, it should be all fixed now in `v0.8.4`!
